src/components/tl/ucp/tl_ucp_sendrecv.h

```diff
@@ -152,8 +154,8 @@ ucc_tl_ucp_recv_common(void *buffer, size_t msglen, ucc_memory_type_t mtype,
     req_param.memory_type = ucc_memtype_to_ucs[mtype];
     req_param.user_data   = (void *)task;
     task->tagged.recv_posted++;
-    return ucp_tag_recv_nbx(UCC_TL_UCP_WORKER(team), buffer, 1, ucp_tag,
-                            ucp_tag_mask, &req_param);
+    return ucp_tag_recv_nbx(UCC_TL_UCP_TEAM_WORKER(team, task), buffer, 1,
```

Comment:
Since service team is a separate ucc team I would rather cached ucp worker on team i.e. service team worker points to service worker or regular worker depending on whether config flag is set or not. Then we can remove is_service from task and remove extra branch from UCC_TL_UCP_TEAM_WORKER

Reply:
I am not sure to understand, what do you mean by "separate ucc team"? The service team is of type `ucc_tl_ucp_team_t` so the only way to point to the worker is first to point to the context which itself stores both ucp_worker and service.ucp_worker.
Unless we duplicate entirely the context for the service team (and replace the regular fields with the one from service), I don't know how to make the "team point to service worker" -- is it what you suggest, to create a special context for the service team?

Reply:
@samnordmann i think what Sergey meant is smth like this:
say we add
```
typedef struct ucc_tl_ucp_worker {
    ucp_context_t ucp_context;
    ucp_worker_t  ucp_workerk;
    ....
    addr
    addrlen
    eps
    ep_hash
} ucc_tl_ucp_worker_t;
```

Then on context:
```
typedef struct ucc_tl_ucp_context {
   ...
    ucc_tl_ucp_worker_t worker;
    ucc_tl_ucp_worker_t service_worker;
  ...
} ucc_tl_ucp_context_t;
```

Team will have a pointer:
```
typedef struct ucc_tl_ucp_team {
   ...
   ucc_tl_ucp_worker_t *worker;
   ...
} ucc_tl_ucp_team_t;
```
Then during team creation:
```
ucc_tl_ucp_team_create_test () {
  ...
  if (TEAM_IS_SERVICE(team) && ctx->cfg.service_worker) {
     team->worker = &ctx->service_worker;
  } else {
     team->worker = &ctx->worker;
  }
}
```

In that case you never need to evaluate "is_service" condition ever again. you always just use team->worker.ucp_worker or team->worker.ep_hash, etc.

Reply:
done, [TL/UCP: add worker struct and ptr in team](https://github.com/openucx/ucc/pull/560/commits/8f45ade0fa9ce5c402214bde92057980b79527cd)

Note that the branching to determine whether the team is the service team still needs to be done and can 't be avoided in the function `ucc_tl_ucp_connect_team_ep` from `src/components/tl/ucp/tl_ucp_ep.c`